homeassistant/components/iotty/switch.py

```diff
@@ -0,0 +1,170 @@
+"""Implement a iotty Light Switch Device."""
+
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from iottycloud.device import Device
+from iottycloud.lightswitch import LightSwitch
+from iottycloud.verbs import LS_DEVICE_TYPE_UID
+
+from homeassistant.components.switch import SwitchDeviceClass, SwitchEntity
+from homeassistant.const import EntityCategory
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import IottyConfigEntry
+from .api import IottyProxy
+from .const import DOMAIN
+from .coordinator import IottyDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: IottyConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Activate the iotty LightSwitch component."""
+    _LOGGER.debug("Setup SWITCH entry id is %s", config_entry.entry_id)
+
+    hass_data = hass.data[DOMAIN]
+
+    coordinator: IottyDataUpdateCoordinator = hass_data[config_entry.entry_id]
+
+    entities = [
+        IottyLightSwitch(
+            coordinator=coordinator, iotty_cloud=coordinator.iotty, iotty_device=d
+        )
+        for d in coordinator.data.devices
+        if d.device_type == LS_DEVICE_TYPE_UID
+    ]
+    _LOGGER.debug("Found %d LightSwitches", len(entities))
+
+    async_add_entities(entities)
+
+    known_devices: set = config_entry.runtime_data.known_devices
+    for known_device in coordinator.data.devices:
+        if known_device.device_type == LS_DEVICE_TYPE_UID:
+            known_devices.add(known_device)
+
+    @callback
+    def async_update_data():
+        """Handle updated data from the API endpoint."""
+        if not coordinator.last_update_success:
+            return None
+
+        devices = coordinator.data.devices
+        entities = []
+        known_devices: set = config_entry.runtime_data.known_devices
+
+        # Add entities for devices which we've not yet seen
+        for device in devices:
+            if (
+                any(d.device_id == device.device_id for d in known_devices)
+                or device.device_type != LS_DEVICE_TYPE_UID
+            ):
+                continue
+
+            entities.extend(
+                [
+                    IottyLightSwitch(
+                        coordinator=coordinator,
+                        iotty_cloud=coordinator.iotty,
+                        iotty_device=device,
+                    )
+                ]
+            )
+            known_devices.add(device)
+
+        async_add_entities(entities)
+
+        return devices
+
+    # Add a subscriber to the coordinator to discover new devices
+    coordinator.async_add_listener(async_update_data)
+
+
+class IottyLightSwitch(SwitchEntity, CoordinatorEntity[Device]):
```

Comment:
Should the switch entity be part of a device instead?

Reply:
Done